### ISSUE TITLE
Performance difference in large Java application between 0.6.0 and 0.5.0 

### ISSUE THREAD

**Issue by jgwest** [eclipse/codewind#1222]
There appears to be a significant performance difference between 0.6.0 and 0.5.0, when tested against a largish Java application (the Rogue Cloud client). 

#### Windows Laptop, local, on Eclipse (Eclipse and Codewind version matching):
```
0.6: consistently about ~40 seconds to build complete, after making a source change (likewise it seems to takes an additional ~5 seconds before the build actually starts, which is not the case on 0.5)
0.5: consistently about ~15  seconds to build complete, after making a source change
```

#### Linux Desktop,  local (desktop machine, but VM on non-SSD) on Eclipse (Eclipse and Codewind version matching):
```
0.6: 40 seconds
0.5: 16 seconds
```

#### Windows Laptop, local, on VSCode (versions matching)
```
VScode on Windows on 0.6: 46 seconds
VSCode on Windows on 0.5: 11* seconds
(* auto build didn't work for unknown reasons, so this is me hitting the 'Build' button, which should be appx the same.)
```

### Steps to run

Run the following (or use the UI in the IDE of your choice):
```
git clone https://github.com/microclimate-dev2ops/rogue-cloud-client-codewind

# WARNING: You MUST rename the project before importing into Codewind, to avoid https://github.com/eclipse/codewind/issues/1217
mv rogue-cloud-client-codewind  gameclient
```

Replace the `gameclient/.cw-settings` file with this:
```
{
  "contextRoot": "",
  "internalPort": "",
  "healthCheck": "",
  "internalDebugPort": "",
  "isHttps": false,
  "ignoredPaths": [
    ""
  ],
  "mavenProfiles": [
    ""
  ],
  "mavenProperties": [
    ""
  ]
}
```
WARNING: If you don't do this, you will get a nasty error message from Eclipse indicating that `watchedFiles` could not be found (this due to https://github.com/eclipse/codewind/issues/133)

Import the project into your workspace, and add it to Codewind. Wait for the project to build and start.

Open StartAgentServlet.java (CTRL-Shift-R to find it in Eclipse, other key combs in other IDEs). Make a change to it. Hit Save.

Immediately start a stopwatch.
Stop the stopwatch when IDE has 'build successful' status on the project (do not use stopping/stopped/starting/started as the indicator to stop the stopwatch.)

### Methodology

This is a holistic approach meant to capture the user's experience of build time:
- Measures the elapsed time between when the user makes a source change, to the UI reporting that the build has completed. 
- Thus it does not use 'starting' state of the server as an indicator (even if we ourselves might know that this could imply the build has completed.)
- We can probably break down the times into additional distinct components: time to see UI first reporting the application is compiling, time to (final) starting status, time for build status reported as success (but breaking down the times is outside the scope of the specific measurement I am doing here)
- Attempt the modify+build operation several times, and time each attempt. Wait for the time to become consistent before recording it (eg avoid the problem of timing the 'cold start')

**Comment by tobespc** [comment]
I'll also investigate this

**Comment by johnmcollier** [comment]
I did some quick testing on Che too:
1. Project import of Rogue cloud is really slow, almost a minute. There's also no import status to let the user know that the project is still importing, making it look like it failed
2. Updates aren't too bad. On Che, we're not doing any docker cp / kubectl cp from PFE (as we're sharing the PFE volume where needed with the project), so the only slowness comes from streaming the file changes from Theia -> PFE.

**Comment by elsony** [comment]
I mark this as stopship for the time being since the original data suggest delta update jumps on Eclipse on Win jumps from 15s->50s.  I am getting the team to run through more scenario (Eclipse vs VSCode, win vs mac, local vs Che) to get more understanding on the extend of the problem. We are also investigating on where the time is spent on.

**Comment by tobespc** [comment]
There are 2 areas where the new design would have an impact on performance,   The first area is where we copy files from the users workspace into the pfe container.  This is done using http upload.   Our testing of this bit shows its very fast.  The second area is where we copy files from the pfe container into the turbine build/run container.  This is done with a docker copy and we run that for each individual file.  I suspect the impact is in the docker copy.  I'm going to put  timing logging into the code at these places to see if it helps with narrowing down the culprit

**Comment by tobespc** [comment]
so we need to understand which bit is the issue.  Is it syncing the users workspace to the pfe container or is it then the update to build/run container

**Comment by johnmcollier** [comment]
Che 0.6.0:
**Import**:
- Total time: 5 min, 50s
  - Sync to PFE: 24s
  - `Build succeeded:` 5 min, 26s
      - Docker build and helm install: 4 min, 52s 
      - Maven build: 34s

**Code update:**
- Total time: 19s
    - Sync to PFE: 12s
    - `Build Succeeded`: 7s

**Comment by tobespc** [comment]
on vscode

Code update 17s in total

**Comment by rajivnathan** [comment]
Tested code update on VSCode on Mac and seeing better times than Jonathan for 0.6.0. Also tried 0.5.0 for comparison and seeing faster times. Perhaps the problems are specific to the eclipse client? Will try eclipse client on my machine in a bit.

**0.6.0**
18s
16s
19s
15s


**0.5.0**
30s
32s
45s
32s

**Comment by tobespc** [comment]
i like those results - 0.6.0 is better :-)

**Comment by maysunfaisal** [comment]
Comparing with John's cluster:

Comparing with John’s:

Che 0.6.0:
*Import*:
- Total time:  4min, 56s
 - Sync to PFE: 23s
 - `Build succeeded:`  4min, 33s
     - Docker build and helm install: 4 min, 07s
     - Maven build: 26s

*Code update:*
- Total time: 17s
   - Sync to PFE: 03s
   - `Build Succeeded`: 14s

**Comment by maysunfaisal** [comment]
But I agree with John's comment regarding import i.e.; if its a big project it takes a while to import like in my case - 23s, user has no way of knowing what is happening during this time. Almost feels like no action is taking place. 

If a user was not following the logs, they'd most probably try importing again!

**Comment by elsony** [comment]
For that usability issue on import with no feedback, I think we should probably open a separate bug on it

**Comment by yangcao77** [comment]
0.6.0 on Mac with vscode

Code update 15s in total. 

- Sync to PFE: 4s

-  Build Succeeded: 11s

**Comment by maysunfaisal** [comment]
I have updated my results with 0.5.0 devfile as well. Looks very comparable.

The only delay seems to be the initial sync of the project during create/import.

**Comment by rtaniwa** [comment]
On W10 and using VS Code I see a pretty significant delay between 0.5 and 0.6. The strange thing is it wasn't auto-building on 0.5, so I had to change the file and initiate the build manually. That build would take about 20s. Now on 0.6, the build does kick off automatically, but does take around 60s.

**Comment by eharris369** [comment]
I can reproduce this problem on my Windows machine.  I ran 3 times each:

**0.6.0:** 47.28, 46.6, 45.49
**0.5.0:** 16.54, 18.04, 17.16

**Comment by elsony** [comment]
It looks like the test result is pretty consistent across the board that it is a generic problem on Win.  Jonathan has also reported the same problem as Rob mentioned that auto build does not kick in.

**Comment by elsony** [comment]
@jgwest Assigning this to you to further investigate on this problem given that you have the environment setup.

**Comment by tobespc** [comment]
I've put a PR up (against master branch) that includes logging information around the timings of each step.  This will make it easier to see what is actually happening with the timings.  Please can you test with this master branch rather than 0.6.0 (It is the same architecture after all) 

https://github.com/eclipse/codewind/pull/1233

**Comment by tobespc** [comment]
in my testing on Windows, I'm seeing bind times of around 4 secs and sync times of around 3 secs for a single file change

**Comment by tobespc** [comment]
@elsony I think it would be worth adding timing information around the build step as well

**Comment by rtaniwa** [comment]
[pfelogs.txt](https://github.com/eclipse/codewind/files/3875794/pfelogs.txt)

**Comment by maysunfaisal** [comment]
On Windows:

[21/11/19 17:57:37 /portal/routes/projects/remoteBind.route.js] [INFO] Total time to bind project gameclient is 6.741 seconds

Total Time from making change to Build Succeeded: 57s
Maven Compile: 15s

[21/11/19 18:08:39 /portal/routes/projects/remoteBind.route.js] [INFO] Total time to sync project gameclient to build container is 3.15 seconds

**Comment by jgwest** [comment]
Here's the PFE log, with trace enabled, with both the Portal and Turbine timing PRs, on Windows, on master branch (PFE master, Eclipse master, CWCTL master):
[cw-pfe-4-trace-with-portal-turbine-timing-prs.txt](https://github.com/eclipse/codewind/files/3876477/cw-pfe-4-trace-with-portal-turbine-timing-prs.txt)

I have annotated the log files with the time the source change was made, and the point at which the build status was reported as completed. Search for `Source code change made in the log.

**Comment by rajivnathan** [comment]
Update: So looks like the problem is on windows we're doing an mvn package instead of compile for some reason but on mac we're doing an mvn compile. Investigating further...

Breakdown from Jonathan's logs:
18:57:04 - file change made
18:57:06 - sync complete
18:57:06 - idc build started
18:57:09 - stopping server
18:57:16 - mvn package start
18:57:31 - mvn package finished
18:57:37 - starting server
18:57:48 - server started
18:57:50 - build complete status sent
Breakdown from my logs on Mac:
19:30:17 - file change made
19:27:57 - sync complete
19:30:19 - idc build started
19:30:22 - mvn compile start
19:30:31 - mvn compile finished
19:30:33 - build complete status sent

**Comment by jgwest** [comment]
With @maysunfaisal's latest PR I have seen the build time drop from ~45 seconds to ~28 seconds, and likewise I confirm we no longer see the `maven package` behaviour that the PR is designed to fix, but it's still a fair way off from the ~15 seconds I was seeing on this machine on `0.5`. 

Here is the the PFE log, on Windows local, with trace enabled, with Maysun's PR, on `0.6` branch:
[fw-pfe-5-trace-with-maysuns-pr.log](https://github.com/eclipse/codewind/files/3877289/fw-pfe-5-trace-with-maysuns-pr.log)

I have annotated the log files with the time the source change was made, and the point at which the build status was reported as completed. Search for `Source code change` in the log.

**Comment by rajivnathan** [comment]
Maysun and I dropped a safe fix (https://github.com/eclipse/codewind/pull/1242) into 0.6.0 for the mvn package problem and verified it improved the performance. We asked Jonathan to try as well but he noticed while the performance is improved it's not quite on par with 0.5.0. He'll post his findings along with the trace logs and we'll have to take a look in the morning.

**Comment by rajivnathan** [comment]
Breakdown from Jonathan's logs after Maysun's fix:

**Build #1**
23:32:06 - /upload called
23:32:08 - sync complete
23:32:08 - idc build started
23:32:10 - mvn compile start
23:32:30 - mvn compile finished
23:32:32 - build complete status sent

Total Time: 26s
mvn compile time from build log file: 16.432 s

**Build #2**
23:35:04 - /upload called
23:35:06 - sync complete
23:35:06 - idc build started
23:35:09 - mvn compile start
23:35:21 - mvn compile finished
23:35:22 - build complete status sent

Total Time: 18s
mvn compile time from build log file: 9.897 s

**Build #3**
23:36:06 - /upload called
23:36:08 - sync complete
23:36:08 - idc build started
23:36:10 - mvn compile start
23:36:22 - mvn compile finished
23:36:24 - build complete status sent

Total Time: 18s
mvn compile time from build log file: 9.535 s

**Build #4**
23:49:00 - Jonathan's marker for Source change made
23:49:04 - /upload called
23:49:06 - sync complete
23:49:06 - idc build started
23:49:08 - mvn compile start
23:49:20 - mvn compile finished
23:49:22 - build complete status sent

Total Time: 18s
Total Time including Jonathan's marker: 22s
mvn compile time from build log file: 9.043 s

Based on the logs it appears that Build 1 was significantly slower at ~26 seconds from the upload timestamp to the build completion status event but builds 2-4 were ~18 seconds which is much closer to the ~15 seconds Jonathan reported on 0.5.0 and the extra sync step for 0.6.0 accounts for the difference.

Based on these results we need to verify 2 things with @jgwest :
1. Is Build 1 the time reported as ~28 seconds in your latest comment? If so, then the subsequent builds show the time is close to the reported times for 0.5.0
2. Is the **Source change made** marker accurate for when the file was saved for Build 4 or would the /upload timestamp be more correct? If the marker is accurate then that means the sync is possibly taking ~4 seconds longer.

I can also try on another windows vm in the morning so we have a different set of results to compare.

**Comment by jgwest** [comment]
@rajivnathan: For consistency, I recommend sticking to the timing methodology described in the original issue, which is _only_ to use the IDE UI to determine time changes, otherwise you are comparing apples (log statement based timing) to oranges (UI behaviour based timing). Using the previously described methodology, I consistently see ~28 seconds with the new `0.6`, and when I last tested ~15 seconds with `0.5` (and see below for the logs for a 18s `0.5` run I got just now). 

Specifically: in the builds you broke down, I hit CTRL-S at exactly :00s (not :06s when the /upload log statement was called), and likewise the UI reported success at appx :28s (not :22 or :24). 

So, for instance, if you believe that the IDE, CWCTL, or FWd is taking longer then they should to call PFE (which produces the `/upload called` log statements), or you believe the IDE is taking too long to report the status change, I recommend investigating those areas.

Here is a trace-enabled log `0.5` on this platform, for further investigation and/or if you would like to explore your alternative timing methodology:
[fw-pfe-6-trace-on-0.5-GA.log](https://github.com/eclipse/codewind/files/3878360/fw-pfe-6-trace-on-0.5-GA.log)

To answer your questions:
1) Build 4 is the time I reported as 28 seconds (and FWIW I recommend against attempting to interpret test results from that logs that I have not explicitly reported results for, eg Builds 1 through 3).
2) Source code change marker is accurate: I make a change a line in a file (but don't save it), wait for the clock to flip to :00, and then hit `CTRL-S`, which should give responses times of <1 second.

**Comment by rajivnathan** [comment]
Yep, agree 100% we're not comparing apples to apples here and your timing methodology is key because that's what a user would see and so should be the timing we focus on for whether we pursue a fix for 0.6.0 or not.

The reason why I asked the two questions was exactly to understand the differences between your reported time and what I see in the logs and to rule out the possibility for discrepancies and also to possibly extrapolate where to look next.

Thanks for the 0.5.0 logs, that's definitely valuable for comparison. Here's what I see from the logs:

07:14:00 - Jonathan's marker for Source change made
n/a (non-hybrid arch) - /upload called
n/a (non-hybrid arch) - sync complete
07:14:02 - idc build started
07:14:04 - mvn compile start
07:14:15 - mvn compile finished
07:14:15 - build complete status sent

Total Time: 15s
mvn compile time from build log file: 8.508 s

Some key takeaways:

Time from Jonathan's marker to the idc build start 
0.5.0 - ~2 seconds
0.6.0 - ~6 seconds

Time from idc build start to build complete status sent
0.5.0 - ~13 seconds
0.6.0 - ~16 seconds

So we'd expect an increase of ~4 seconds with the new architecture and there is an unexpected increase of ~3 seconds in the build time but we're seeing a difference of ~13 seconds in reality which means there's ~6 seconds unaccounted for.

**Comment by elsony** [comment]
Based on the discussion on the stopship call today, we have improved on the latest fix but it is still longer than we want. However, based on the current data, it doesn't looks like the remaining time will be easy enough for us to drastically improve on 0.6.0.  Therefore, we are moving the remaining performance improvement work to 0.7.0.

**Comment by malincoln** [comment]
@ssh24 confirming that the fix is on track for 0.7.0?

**Comment by ssh24** [comment]
I don't think I was assigned on this task for any particular fix. I was more of assigned to https://github.com/eclipse/codewind/issues/1245 to find the benchmark times across the projects in the turbine module

**Comment by elsony** [comment]
@malincoln For this one, the major part of the original problem has been fixed before we shipped 0.6.0.  We are more in the process of doing investigation to see if we can figure out where that extra 6 sec is spent.  Unless we can find something early, I don't think we'll explicitly be doing anything for 0.7.0 on this.

**Comment by malincoln** [comment]
@elsony noted and thanks for the clarification.

**Comment by rajivnathan** [comment]
Discussed this in the iterative-dev status meeting today. We'll run tests with 0.8 for comparison with previous results and schedule a weekly run of the performance tests.

**Comment by ssh24** [comment]
Something I noticed was that the cwctl installation in our test script was always pointing to master and for example when trying to do performance test on 0.8.0 it would fail on cwctl create project command as the syntax has changed in master. I have opened https://github.com/eclipse/codewind/pull/1868 to fix that. Once that lands in master, we should also backport it to 0.8.0 and 0.7.0 in order to get the metrics from those releases.

**Comment by ssh24** [comment]
Here is the 0.8.0 data:
![Screen Shot 2020-01-28 at 10 36 24 AM](https://user-images.githubusercontent.com/15173354/73279469-2d204200-41bb-11ea-9dba-a022cfa3f481.png)

Looks like the performance test changes were not put into 0.7.0 so when tried running against 0.7.0 it didn't yield any result.

**Comment by ssh24** [comment]
Here is the 0.7.0 data:
![Screen Shot 2020-01-30 at 9 24 32 AM](https://user-images.githubusercontent.com/15173354/73458450-494eeb00-4343-11ea-93f0-dd860b795abc.png)

Looks like liberty failed to edit dockerfile or even example.java script throughout all iterations and that's why there is no time there. Looking at the logs it looks like the issue we were seeing in Liberty and the one @maysunfaisal fixed recently on master. Since it was not backported to 0.7.0 it still fails in 0.7.0

**Comment by ssh24** [comment]
Here is the latest data after fixing the 20s issue on our test automation for update from master:

![Screen Shot 2020-02-04 at 10 04 44 AM](https://user-images.githubusercontent.com/15173354/73756566-c8726380-4735-11ea-8f90-d99df6256793.png)

**Comment by rajivnathan** [comment]
As discussed in the previous meeting it would be good if we could start tracking differences from the previous release to master and flag performance regressions in an automated way.

**Comment by ssh24** [comment]
Here is the result from last night:
![Screen Shot 2020-02-06 at 2 07 35 PM](https://user-images.githubusercontent.com/15173354/73969824-17f99080-48ea-11ea-9b74-0581db7c6e67.png)

I am thinking this story can be closed (has been open for a while) and then we can create another story that handles creating a cron that runs every night and a cron that runs after every DCUT with probably a webserver page that posts the result and comparisons of the last two runs.

For example: for master runs it would compare the last two runs to see if any time increased, for dcut runs it would compare with the last master run.

Thoughts?

**Comment by ssh24** [comment]
Issue for performance test cron: https://github.com/eclipse/codewind/issues/2091

**Comment by ssh24** [comment]
0.9.0 data:
![Screen Shot 2020-02-12 at 2 32 25 PM](https://user-images.githubusercontent.com/15173354/74370135-951e7d00-4da4-11ea-8948-6b92b258809e.png)

**Comment by ssh24** [comment]
Here is the data from last night's kube performance run from master. Odo projects failed on create so it has no metrics:
<img width="1213" alt="Screen Shot 2020-02-18 at 9 13 35 AM" src="https://user-images.githubusercontent.com/15173354/74743817-135ca280-522f-11ea-8777-4c8b556b9dfb.png">

**Comment by ssh24** [comment]
So I have been doing some investigation on this issue on the basis of testing out a large java application (i.e gameclient) on Mac and Windows across three different release: 0.5.0, 0.6.0 and 0.9.0 and checking the time differences across docker build and maven build.
1. I couldn't replicate the numbers Jonathan have found on his windows run after trying multiple times. I was however able to replicate the numbers Stephanie/Rajiv found on mac repeatedly. I have put on a comparison table that compares build times across 0.5.0, 0.6.0 and 0.9.0 on windows and mac. It is inevitable from the graph, our build time has become better after 0.5.0 release. Please see the attach excel spreadsheet for the comparison.
[0.5.0-0.6.0-0.9.0.xlsx](https://github.com/eclipse/codewind/files/4325757/0.5.0-0.6.0-0.9.0.xlsx)
2. Another fact I noticed is that during maven build you will notice:
```
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ gameclient ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 161 source files to /home/default/app/mc-target/classes
```
The compilation of 161 source file was interesting but it was found that maven will compile all the files in the module on a single source change.
3. Another thing we found was that during maven build if the actual maven compile takes X amount of seconds, the overall build (i.e from event Compiling Application to even Build Succeeded) is actually some positive offset of X (i.e X + V + Y where V is a variable time for the overall maven compile command to finish and Y is a variable time where turbine does a little bit of work to call the logs api and return the log file content). Example:
- **The compiling application event was shown @ 18:09:22**
```
[12/03/20 18:09:22 gameclient] [INFO] Emitting event
 message: projectStatusChanged
 data: {
  "projectID": "4730dcf0-6474-11ea-af84-79cbb6843c43",
  "buildStatus": "inProgress",
  "detailedBuildStatus": "Compiling application"
}
```
- **Application build starts @ 18:09:22**
```
Starting mvn compile Thu Mar 12 18:09:22 UTC 2020
```
- **Maven build finished @ 18:09:32**
```
Finished mvn compile for cw-4730dcf0-6474-11ea-af84-79cbb6843c43-81de77ba107bb4ab7f93f3f8a0f551eb122c2a90 Thu Mar 12 18:09:32 UTC 2020
```
- At this point one would argue the overall maven build was ~10s
- However, when looking at the maven build log
```
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ gameclient ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 161 source files to /home/default/app/mc-target/classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.917 s
[INFO] Finished at: 2020-03-12T18:09:32Z
[INFO] ------------------------------------------------------------------------
```
- The `mvn compile` log actually took ~9.7s
```
executeBuildScript : 385 |
real	0m9.770s
```
- So somehow, maven is taking those 4 extra seconds and we have no control over them
- **The project succeeded event was shown @ 18:09:34**
```
[12/03/20 18:09:34 gameclient] [INFO] Emitting event
 message: projectStatusChanged
 data: {
  "projectID": "4730dcf0-6474-11ea-af84-79cbb6843c43",
  "buildStatus": "success",
  "detailedBuildStatus": " ",
  "lastbuild": 1584036574842
}
```
- Using the mechanism to compare the timing by looking at the IDE as the events changes, it was depicted the overall maven build was ~12s

Therefore, the breakdown:
- 6s for actual maven build
- 4s extra for the actual maven compile to return back (in total 10s of mvn compile command to finish when the actual maven build was 6s)
- 2s for turbine to do rest of the stuff i.e call the logs api

These were the findings for this issue - **performance has actually improved from 0.5.0 to 0.6.0 with the use of sync mechanism**. With respect to the mvn compile taking longer to return back to us, not sure if we can do anything to mitigate that.

We also now have automated performance testing that runs nightly against master branch to pick up on any deterioration of the turbine code base. It also runs whenever we have a release against an older release and reports back any anomalies in project creation, update or delete.  See more about it [here](https://github.com/eclipse/codewind/tree/master/src/pfe/file-watcher/server/test#performance-test-suite)

**Comment by ssh24** [comment]
Based on the discussion above, the turbine team has decided that we cannot do much about the maven compile command process taking longer to return back to us.

Closing this issue as resolved.

/close
